Update best_header inside Connect/DisconnectTip

aureleoules opened the pull request:
Fixes #26245. 
From furszy:
> Update best_header inside Connect/DisconnectTip
> This solves a problem where best_header isn't updated in case of forced reorg
> like a block invalidation and reconsideration.

I added a test.

aureleoules committed:
test: Check reconsiderblock sets best header

dergoegge reviewed (changes requested):
I think this is a validation bug not a rpc bug, so setting `m_best_header` (validation specific state) on an rpc call does not seem right to me.

Currently, we only ever set `m_best_header` in three places:

When we add a new header to our index:

https://github.com/bitcoin/bitcoin/blob/5e82b9ba96b6c5614a1187382a086e5694dff544/src/validation.cpp#L3676

When an invalid block is found:

https://github.com/bitcoin/bitcoin/blob/5e82b9ba96b6c5614a1187382a086e5694dff544/src/validation.cpp#L1606

Or upon startup while loading the block index into memory:

https://github.com/bitcoin/bitcoin/blob/5e82b9ba96b6c5614a1187382a086e5694dff544/src/validation.cpp#L4268


afaict anytime we are reorging to an already known chain (i.e. we already have the headers in our index) we will not set `m_best_headers` until a new header is added to our index on that chain.

aureleoules commented:
Thanks @dergoegge and @Sjors. I think you are correct, I've moved the statement inside `ActivateBestChain` in `validation.cpp`. Is it better?
I also added your test vector @dergoegge.

Sjors commented:
It seems like a better place. I'll have to think more about the change itself of course. Paging @ryanofsky and @jamesob who hang out in this part of the consensus code more often...

furszy commented on src/validation.cpp:
```diff
@@ -3018,6 +3018,7 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
                     pindexMostWork = nullptr;
                 }
                 pindexNewTip = m_chain.Tip();
+                m_chainman.m_best_header = pindexNewTip;
```
Pretty sure that this isn't correct, the headers-only chain is synchronized first. So, during IBD, it will be ahead of the "tip" block that we are connecting, and you are going to overwrite it.

dergoegge replied:
```suggestion
                if(pindexNewTip->nChainWork > m_chainman.m_best_header->nChainWork) {
                    m_chainman.m_best_header = pindexNewTip;
                }
```

furszy committed:
validation: bugfix, update best_header inside Connect/DisconnectTip

This solves a problem where best_header isn't updated in case of forced reorg
like a block invalidation and reconsideration.

furszy commented:
Quick thing: in case of regular headers-sync, we only reorg if the chain has a higher diff, so `best_header` field is fine there.

In case of forced reorgs, like a block invalidation + reconsideration, we aren’t reorganizing the chain due PoW, thus why the `best_header` isn’t updated (we currently only update the `best_header` field after adding it to the block index, inside `AddToBlockIndex`).

So, for those cases, could either update `best_header` inside `ConnectTip` and `DisconnectTip`  like https://github.com/furszy/bitcoin-core/commit/52f5d62957ad0d9e2b2c8d00333a369fb4a369c7 or.. update it inside the reconsider block process only (the `invalidate block process is fine because we are updating the best_header inside `InvalidChainFound`).

aureleoules commented:
Thanks @furszy, I've cherry-picked your commit.

luke-jr commented on src/validation.cpp:
```diff
@@ -2632,6 +2632,12 @@ bool Chainstate::DisconnectTip(BlockValidationState& state, DisconnectedBlockTra
 
+    // if we are roll-backing the chain, rollback the best header as well
+    if (m_chainman.m_best_header && m_chainman.m_best_header->GetBlockHash() == pindexDelete->GetBlockHash()
+        && pindexDelete->pprev) {
```
Why the pprev check? If we really need that, maybe an assert would be better?

luke-jr commented on src/validation.cpp:
```diff
@@ -2745,6 +2751,12 @@ bool Chainstate::ConnectTip(BlockValidationState& state, CBlockIndex* pindexNew,
     m_chain.SetTip(*pindexNew);
+
+    // If the best header is the prev block, update the best header
+    if (pindexNew->pprev && m_chainman.m_best_header && m_chainman.m_best_header->GetBlockHash() == pindexNew->pprev->GetBlockHash()) {
```
```c++
    if (pindexNew->pprev == m_chainman.m_best_header) {
```
?

luke-jr reviewed (changes requested):
This feels like a special-cased hack. Is there a reason we can't adjust the current code that changes m_best_header when an invalid block is encountered?

(Also, PR title needs updating.)

aureleoules commented:
Thanks @luke-jr for the review.
> This feels like a special-cased hack. Is there a reason we can't adjust the current code that changes m_best_header when an invalid block is encountered?

I believe it is more robust to place the fix in ConnectTip and DisconnectTip. cc @furszy

furszy commented:
> This feels like a special-cased hack. Is there a reason we can't adjust the current code that changes m_best_header when an invalid block is encountered?

I actually wrote both ways of fixing this above https://github.com/bitcoin/bitcoin/pull/26260#issuecomment-1270121139.

The problem is on the block reconsideration flow, not inside the invalidation one.

Inside the `invalidateblock` process, we update the best_header by
manually calling `InvalidChainFound` after invalidating blocks. Which is fine.

We need to do the same for `reconsiderblock` and update the
best_header field after finishing the process.

Note: the only difference between this two commands is that
‘reconsiderblock’ does not have its own separate function,
it's a plain RPC command that resets the block index flag
and calls ActivateBestChain.

So, @aureleoules, could drop the other commit and cherry-pick this change to fix the problem too:
https://github.com/furszy/bitcoin-core/commit/4f3318c684af631d7f79b872cd7b8d4feae4b828

Both works. This last one is specific to the `reconsiderblock` flow (which is where we have the issue).

Sjors commented:
I deployed this patch on the two ForkMonitor (mirror) nodes that frequently use `invalidateblock` and `reconsiderblock`. It's cherry-picked on top of the v25.0rc1 and v24.1rc2 tags. Typically we get about one error message per week where the block it rewinds to is not the one we expected. We'll see if that stops.

pinheadmz commented on src/validation.cpp:
```diff
@@ -2632,6 +2632,12 @@ bool Chainstate::DisconnectTip(BlockValidationState& state, DisconnectedBlockTra
     m_chain.SetTip(*pindexDelete->pprev);
 
+    // if we are roll-backing the chain, rollback the best header as well
+    if (m_chainman.m_best_header && m_chainman.m_best_header->GetBlockHash() == pindexDelete->GetBlockHash()) {
+        assert(pindexDelete->pprev);
+        m_chainman.m_best_header = pindexDelete->pprev;
+    }
```
Tests still pass with this chunk removed. I wonder if checking that getblockchaininfo `headers == blocks` isn't enough? Might need to use `getchaintips` to cover.

pinheadmz reviewed (changes requested):
concept ACK

one question below.

pinheadmz commented on test/functional/rpc_invalidateblock.py:
```diff
@@ -20,6 +20,10 @@ def set_test_params(self):
         self.setup_nodes()
 
+    def check_blocks_headers(self, node):
+        blockchaininfos = node.getblockchaininfo()
```
micro-nit, this is a silly variable name. i think it's fine, but it is silly.

achow101 commented:
Closing as up for grabs due to lack of activity.
